File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/329-longest-increasing-path-in-a-matrix-approach3.go
package main

import "fmt"

func longestIncreasingPath(matrix [][]int) int {
	if len(matrix) == 0 {
		return 0
	}
	rows, cols := len(matrix), len(matrix[0])
	inDeg := make([][]int, rows)
	for i := range inDeg {
		inDeg[i] = make([]int, cols)
	}
	dirs := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			for _, d := range dirs {
				nr, nc := r+d[0], c+d[1]
				if nr >= 0 && nr < rows && nc >= 0 && nc < cols && matrix[nr][nc] < matrix[r][c] {
					inDeg[r][c]++
				}
			}
		}
	}
	queue := [][2]int{}
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if inDeg[r][c] == 0 {
				queue = append(queue, [2]int{r, c})
			}
		}
	}
	levels := 0
	qi := 0
	for qi < len(queue) {
		levels++
		size := len(queue) - qi
		for s := 0; s < size; s++ {
			r, c := queue[qi][0], queue[qi][1]
			qi++
			for _, d := range dirs {
				nr, nc := r+d[0], c+d[1]
				if nr >= 0 && nr < rows && nc >= 0 && nc < cols && matrix[nr][nc] > matrix[r][c] {
					inDeg[nr][nc]--
					if inDeg[nr][nc] == 0 {
						queue = append(queue, [2]int{nr, nc})
					}
				}
			}
		}
	}
	return levels
}

func main() {
	assert := func(cond bool) {
		if !cond {
			panic("assertion failed")
		}
	}
	assert(longestIncreasingPath([][]int{{9, 9, 4}, {6, 6, 8}, {2, 1, 1}}) == 4)
	assert(longestIncreasingPath([][]int{{3, 4, 5}, {3, 2, 6}, {2, 2, 1}}) == 4)
	assert(longestIncreasingPath([][]int{{1}}) == 1)
	assert(longestIncreasingPath([][]int{{1, 1}, {1, 1}}) == 1)
	assert(longestIncreasingPath([][]int{{1, 2, 3, 4}}) == 4)
	fmt.Println("all tests pass")
}

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/329-longest-increasing-path-in-a-matrix-approach3.py
from collections import deque

def longest_increasing_path(matrix: list[list[int]]) -> int:
    if not matrix:
        return 0
    rows, cols = len(matrix), len(matrix[0])
    in_deg = [[0] * cols for _ in range(rows)]
    dirs = ((1, 0), (-1, 0), (0, 1), (0, -1))
    for r in range(rows):
        for c in range(cols):
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and matrix[nr][nc] < matrix[r][c]:
                    in_deg[r][c] += 1
    q = deque()
    for r in range(rows):
        for c in range(cols):
            if in_deg[r][c] == 0:
                q.append((r, c))
    levels = 0
    while q:
        levels += 1
        for _ in range(len(q)):
            r, c = q.popleft()
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and matrix[nr][nc] > matrix[r][c]:
                    in_deg[nr][nc] -= 1
                    if in_deg[nr][nc] == 0:
                        q.append((nr, nc))
    return levels

assert longest_increasing_path([[9, 9, 4], [6, 6, 8], [2, 1, 1]]) == 4
assert longest_increasing_path([[3, 4, 5], [3, 2, 6], [2, 2, 1]]) == 4
assert longest_increasing_path([[1]]) == 1
assert longest_increasing_path([[1, 1], [1, 1]]) == 1
assert longest_increasing_path([[1, 2, 3, 4]]) == 4
print("all tests pass")
